### PR TITLE
feature: run a clean-up job when uninstalling

### LINKED ISSUE #2653: run a clean-up job when uninstalling
### What would you like to be added?

run a clean-up job when uninstalling , so avoid to call clean-up.sh

job
annotations:
    "helm.sh/hook": pre-delete
    "helm.sh/hook-delete-policy": hook-succeeded,hook-failed

### Why is this needed?

_No response_

### How to implement it (if possible)?

d

### Additional context

_No response_

### I've read and agree with the following

- [X] I've checked all open and closed issues and my request is not there.
- [X] I've checked all open and closed pull requests and my request is not there.

### PR DESCRIPTION
**Pull request by ty-dc**
#### What type of PR is this?

**What this PR does / why we need it**:
close #2653

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

  1. Remove all spiderpool mutatingwebhookconfigurations
  2. Remove all spiderpool validatingwebhookconfigurations
  3. Remove all spiderpool CR：SpiderIPPool，SpiderSubnet、SpiderEndpoint, etc.

### CODE REVIEW

**Comment by weizhoublue** [comment]
could it be tested in a E2E case ?

File: cmd/spiderpool-controller/cmd/clean.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+		for _, item := range ipPoolList.Items {
+			item.Finalizers = make([]string, 0)
+			err := c.Update(ctx, &item)
+			if err != nil {
+				return err
+			}
+
+			err = c.Delete(ctx, &item)
+			if err != nil {
+				return err
```
Comment (weizhoublue) [inline cmd/spiderpool-controller/cmd/clean.go:98]: for cleaning job, it should not be interrupted, and clean as much as possible

File: cmd/spiderpool-controller/cmd/clean.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+		err := c.Delete(ctx, vObj)
+		if err != nil {
+			return err
+		}
+	}
+
+	mObj := &webhook.MutatingWebhookConfiguration{}
+	err = c.Get(ctx, client.ObjectKey{Name: mutating}, mObj)
+	if err == nil {
+		err := c.Delete(ctx, mObj)
```
Comment (weizhoublue) [inline cmd/spiderpool-controller/cmd/clean.go:80]: there should be some log for what it succeeded to clean

**Comment by ty-dc** [comment]
```
Run codecov/codecov-action@v3.1.4
  with:
    directory: ./
    files: coverage.out
    flags: unittests
    name: my-codecov-umbrella
    fail_ci_if_error: true
    verbose: true
```

should look like this:
```
Run codecov/codecov-action@v3.1.4
  with:
    directory: ./
    files: coverage.out
    flags: unittests
    name: my-codecov-umbrella
    fail_ci_if_error: true
    verbose: true
    token: ***
```

File: cmd/spiderpool-controller/cmd/clean.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+			logger.Sugar().Errorf("failed to delete MutatingWebhookConfiguration: %v.", err)
+		}
+		logger.Sugar().Infof("Successfully deleted MutatingWebhookConfiguration")
+	}
+
+	ipPoolList := new(spiderpoolv2beta1.SpiderIPPoolList)
+	err = c.List(ctx, ipPoolList)
+	if err == nil {
+		for _, item := range ipPoolList.Items {
+			item.Finalizers = make([]string, 0)
```
Comment (weizhoublue) [inline cmd/spiderpool-controller/cmd/clean.go:92]: log

File: cmd/spiderpool-controller/cmd/clean.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+				logger.Sugar().Errorf("failed to delete ippool's: %v, %v ", &item, err)
+			}
+		}
+	}
+
+	subnetList := new(spiderpoolv2beta1.SpiderSubnetList)
+	err = c.List(ctx, subnetList)
+	if err == nil {
+		for _, item := range subnetList.Items {
+			err = c.Delete(ctx, &item)
```
Comment (weizhoublue) [inline cmd/spiderpool-controller/cmd/clean.go:109]: log

File: cmd/spiderpool-controller/cmd/clean.go [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+				logger.Sugar().Errorf("failed to delete subnet's: %v, %v ", &item, err)
+			}
+		}
+	}
+
+	spiderEndpointList := new(spiderpoolv2beta1.SpiderEndpointList)
+	err = c.List(ctx, spiderEndpointList)
+	if err == nil {
+		for _, item := range spiderEndpointList.Items {
+			item.Finalizers = make([]string, 0)
```
Comment (weizhoublue) [inline cmd/spiderpool-controller/cmd/clean.go:120]: log

File: cmd/spiderpool-controller/cmd/clean.go [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+				logger.Sugar().Errorf("failed to delete spiderEndpoint's: %v, %v ", &item, err)
+			}
+		}
+	}
+
+	reservedIPList := new(spiderpoolv2beta1.SpiderReservedIPList)
+	err = c.List(ctx, reservedIPList)
+	if err == nil {
+		for _, item := range reservedIPList.Items {
+			err = c.Delete(ctx, &item)
```
Comment (weizhoublue) [inline cmd/spiderpool-controller/cmd/clean.go:137]: log

File: cmd/spiderpool-controller/cmd/clean.go [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+			if err != nil {
+				logger.Sugar().Errorf("failed to delete spiderReservedIP's: %v, %v ", &item, err)
+			}
+		}
+	}
+
+	return nil
```
Comment (weizhoublue) [inline cmd/spiderpool-controller/cmd/clean.go:144]: 中间有哪个清除失败了，最终返回错误，反应在 job 的 状态中

File: cmd/spiderpool-controller/cmd/clean.go [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+	spiderEndpointList := new(spiderpoolv2beta1.SpiderEndpointList)
+	err = c.List(ctx, spiderEndpointList)
+	if err == nil {
+		for _, item := range spiderEndpointList.Items {
+			item.Finalizers = make([]string, 0)
+			err := c.Update(ctx, &item)
+			if err != nil {
+				return fmt.Errorf("failed to update spiderEndpoint's: %v finalizers to be empty: %v ", &item, err)
+			}
+			logger.Sugar().Infof("succeeded to update spiderEndpoint's: %v finalizers to be empty", &item)
```
Comment (weizhoublue) [inline cmd/spiderpool-controller/cmd/clean.go:128]: succeeded to clean the finalizers of spiderEndpoint xxxx

File: charts/spiderpool/values.yaml [hunk 9]
```diff
@@ -757,6 +757,10 @@ spiderpoolController:
       ## @param spiderpoolController.tls.auto.extraDnsNames extra DNS names of server cert for auto method
       extraDnsNames: []
 
+  cleanup:
```
Comment (weizhoublue) [inline charts/spiderpool/values.yaml:760]: mention it in the document of uninstalling

File: cmd/spiderpool-controller/cmd/clean.go (hunk 7 above)
Comment (weizhoublue) [reply cmd/spiderpool-controller/cmd/clean.go:150]: +

**Comment by weizhoublue** [comment]
@cyclinder 现在垃圾资源的清除，除了 Spiderpool 本身，其它 项目 的垃圾资源 是否还需要 帮助删除 ？

File: docs/usage/install/uninstall.md [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+2. Uninstall Spiderpool via `helm uninstall`
+
+     ```bash
+     helm uninstall <spiderpool-name> --namespace <spiderpool-namespace>
+     ```
+
+     Replace `<spiderpool-name>` with the name of the Spiderpool you want to uninstall and `<spiderpool-namespace>` with the namespace of the Spiderpool.
+
+### In versions higher than 0.9.2
```
Comment (windsonsea) [inline docs/usage/install/uninstall.md:29]: Above v0.9.2

File: docs/usage/install/uninstall.md [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+```bash
+kubectl get spidersubnets.spiderpool.spidernet.io -o name | wc -l
+kubectl get spiderips.spiderpool.spidernet.io -o name | wc -l
+kubectl get spiderippools.spiderpool.spidernet.io -o name | wc -l
+kubectl get spiderreservedips.spiderpool.spidernet.io -o name | wc -l
+kubectl get spiderendpoints.spiderpool.spidernet.io -o name | wc -l
+kubectl get spidercoordinators.spiderpool.spidernet.io -o name | wc -l
+```
+
+### In versions lower than 0.9.2
```
Comment (windsonsea) [inline docs/usage/install/uninstall.md:42]:
```suggestion
### Below v0.9.2
```

File: docs/usage/install/uninstall-zh_CN.md [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+2. 通过 `helm  uninstall` 卸载 Spiderpool
+
+    ```bash
+    helm uninstall <spiderpool-name> --namespace <spiderpool-namespace>
+    ```
+
+    将 `<spiderpool-name>` 替换为要卸载的 Spiderpool 的名称，将 `<spiderpool-namespace>` 替换为 Spiderpool 所在的命名空间。
+
+### 在高于 0.9.2 的版本
```
Comment (windsonsea) [inline docs/usage/install/uninstall-zh_CN.md:29]:
```suggestion
### 0.9.2 以上版本
```

File: docs/usage/install/uninstall-zh_CN.md [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+```bash
+kubectl get spidersubnets.spiderpool.spidernet.io -o name | wc -l 
+kubectl get spiderips.spiderpool.spidernet.io -o name | wc -l
+kubectl get spiderippools.spiderpool.spidernet.io -o name | wc -l
+kubectl get spiderreservedips.spiderpool.spidernet.io -o name | wc -l
+kubectl get spiderendpoints.spiderpool.spidernet.io -o name | wc -l
+kubectl get spidercoordinators.spiderpool.spidernet.io -o name | wc -l
+```
+
+### 在低于 0.9.2 的版本
```
Comment (windsonsea) [inline docs/usage/install/uninstall-zh_CN.md:42]:
```suggestion
### 0.9.2 以下版本
```

**Review by windsonsea — COMMENTED** [review]
Headings can be shorter and better

**Review by cyclinder — COMMENTED** [review]
Can you update your PR description to list all the resources you clean? I think we should clean up the below resources in order:

1. All the CRD instances: IPPools,Subnets,SpiderMultusConfig...
2. The CRD definitions.
3. Pods that have not been completely deleted, Example sriov-network-config-daemon, see https://github.com/spidernet-io/spiderpool/issues/2822

**Comment by ty-dc** [comment]
> Can you update your PR description to list all the resources you clean? I think we should clean up the below resources in order:
> 
> 1. All the CRD instances: IPPools,Subnets,SpiderMultusConfig...
> 2. The CRD definitions. CRD 定义。
> 3. Pods that have not been completely deleted, Example sriov-network-config-daemon, see [clean sriov pod #2822](https://github.com/spidernet-io/spiderpool/issues/2822)

updated，Step 2 Do it happily. But step 3, what if the customer installed it themselves? Is there any way we can know whether a resource like sirov was installed via spiderpool?

File: docs/usage/install/uninstall-zh_CN.md [hunk 14]
```diff
@@ -0,0 +1,44 @@
+# 卸载指南
```
Comment (weizhoublue) [inline docs/usage/install/uninstall-zh_CN.md:1]: 存量文档 还有很多关于 uninstall 说明
需要链接过来

File: docs/usage/install/uninstall-zh_CN.md [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+    ```bash
+    helm uninstall <spiderpool-name> --namespace <spiderpool-namespace>
+    ```
+
+    将 `<spiderpool-name>` 替换为要卸载的 Spiderpool 的名称，将 `<spiderpool-namespace>` 替换为 Spiderpool 所在的命名空间。
+
+### 0.9.2 以上版本
+
+在 0.9.2 之后引入了自动清理 Spiderpool 资源的功能，它通过 `spiderpoolController.cleanup.enabled` 配置项来启用，该值默认为 `true`，您可以通过如下方式验证与 Spiderpool 相关的资源数量是否自动被清理。
```
Comment (weizhoublue) [inline docs/usage/install/uninstall-zh_CN.md:31]: 为什么是 0.9.2 ， 而不是 下一个 1.0 什么的版本

**Review by weizhoublue — APPROVED** [review]

**Pull request merged**